modules/metrics/README.md

```diff
@@ -0,0 +1,175 @@
+|-------------|----------------------------|------------------------------------------------------|
+| what        | "endpoint-request-duration"|                                                      |
+
+### Fan-out
+
+A Histogram metric, tagged with:
+
+| tag         | value                      | comment                                              |
+|-------------|----------------------------|------------------------------------------------------|
+| what        | "request-fanout-factor"    |                                                      |
+| unit        | "request/request"          |                                                      |
+
+This metric will show you how many downstream requests each endpoint tends to make.
+
+
+## Custom Metrics
+
+To set up custom metrics on the application level you'll need to get hold of the
+`SemanticMetricRegistry` object. Use:
+```java
+  environment.resolve(SemanticMetricRegistry.class);
+```
+
+You can then construct `MetricId`:s and emit metrics in handlers and middleware.
+
+To tag custom metrics with endpoint information, you could pass in the endpoint name, or transform
+whole routes. The example below shows both approaches.
+
+Alternatively, you could set up an `EndpointRunnableFactoryDecorator` and register
+it, similar to how the metrics module does:
+```java
+    Multibinder.newSetBinder(binder(), EndpointRunnableFactoryDecorator.class)
+        .addBinding().to(MetricsCollectingEndpointRunnableFactoryDecorator.class);
+```
+See [Extending incoming/outgoing request handling]
+(https://github.com/spotify/apollo/tree/master/apollo-environment#extending-incomingoutgoing-request-handling)
+for some description of how to do request handling decorations.
+
+For client-side metrics, wrap the ```Client``` you get from the ```RequestContext``` in a similar
+way to how ```DecoratingClient``` is implemented. In your wrapper, ensure that the right metrics
+are tracked.
+
+### Custom per-endpoint response payload size histogram example
+
+```java
+  /**
+   * Use this method to transform a route to one that tracks response payload sizes in a Histogram,
+   * tagged with an endpoint tag set to method:uri of the route.
+   */
+  public Route<AsyncHandler<Response<ByteString>>> withResponsePayloadSizeHistogram(
+      Route<AsyncHandler<Response<ByteString>>> route) {
+
+    String endpointName = route.method() + ":" + route.uri();
+
+    return route.withMiddleware(responsePayloadSizeHistogram(endpointName));
+  }
+
+  /**
+   * Middleware to track response payload size in a Histogram,
+   * tagged with an endpoint tag set to the given endpoint name.
+   */
+  public Middleware<AsyncHandler<Response<ByteString>>, AsyncHandler<Response<ByteString>>>
+      responsePayloadSizeHistogram(String endpointName) {
+
+    final MetricId histogramId = MetricId.build()
+        .tagged("service", serviceName)
+        .tagged("endpoint", endpointName)
+        .tagged("what", "endpoint-response-size");
```

Comment:
This metric name could easily be misconstrued as the total response size. Maybe call it `endpoint-response-payload-size`?